Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -323,48 +330,67 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
         [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
         private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
         {
+            ClientResetHandlerBase.BeforeResetCallback cb = null;
+            SyncConfigurationBase syncConfig = null;
+
             try
             {
                 var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
-                var syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+                syncConfig = (SyncConfigurationBase)syncConfigHandle.Target;
+                cb = syncConfig.ClientResetHandler switch
+                {
+                    DiscardUnsyncedChangesHandler handler => handler.OnBeforeReset,
+                    RecoverUnsyncedChangesHandler handler => handler.OnBeforeReset,
+                    RecoverOrDiscardUnsyncedChangesHandler handler => handler.OnBeforeReset,
+                    _ => throw new NotSupportedException($"ClientResetHandlerBase of type {syncConfig.ClientResetHandler.GetType()} is not handled yet")
+                };
 
-                if (syncConfig.ClientResetHandler is DiscardLocalResetHandler discardLocalHandler &&
-                    discardLocalHandler.OnBeforeReset != null)
+                if (cb != null)
                 {
                     var schema = syncConfig.GetSchema();
                     using var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfig, schema);
-                    discardLocalHandler.OnBeforeReset(realmBefore);
+                    cb.Invoke(realmBefore);
                 }
             }
             catch (Exception ex)
             {
-                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing DiscardLocalResetHandler.OnBeforeReset during a client reset: {ex}");
+                Logger.Default.Log(LogLevel.Error, $"An error has occurred while executing {syncConfig.ClientResetHandler.GetType().Name}.{cb.GetType().Name} during a client reset: {ex}");
```

Comment:
Let's call this `OnBeforeReset` - there's no need to check the type of the callback.

Reply:
All this was done to give you e.g. `DiscardUnsyncedChangesHandler.OnBeforeReset`, which I checked and it works as expected.

Reply:
I'm not sure how it's possible to give you `DiscardUnsyncedChangesHandler.OnBeforeReset`. I haven't tried it, but I expect `cb` here to be of type `BeforeResetCallback` or `Action<Realm>`. There's no obvious reason/way for it to ever be of type `OnBeforeReset`, considering that's the name of a property and not an actual type.

Reply:
What I said makes no sense. Without reflection this is impossible possible. And yet my memory told me "yes yes it's fine". Eh...🤦🏻